tests/unit/s2n_self_talk_threads_test.c

```diff
@@ -0,0 +1,199 @@
+
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    int bytes_written = 0;
+    while (bytes_written < sizeof(data_to_send)) {
+        int w = s2n_send(conn, data_to_send + bytes_written, sizeof(data_to_send) - bytes_written, &blocked);
+        if (w >= 0) {
+            bytes_written += w;
+        } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            printf("Error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return &test_failure;
+        }
+    }
+
+    if (conn->actual_protocol_version < S2N_TLS13) {
+        /* Only TLS1.3 supports half-close */
+        return &test_success;
+    }
+
+    while (s2n_shutdown_send(conn, &blocked) != S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            printf("Error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return &test_failure;
+        }
+    }
+    return &test_success;
+}
+
+/* Should match the s2n_recv example in USAGE-GUIDE.md */
+static void *s2n_test_recv(void *input)
+{
+    struct s2n_connection *conn = (struct s2n_connection *) input;
+
+    char buffer[sizeof(data_to_send)] = { 0 };
+
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    int bytes_read = 0;
+    while (bytes_read < sizeof(buffer)) {
+        int r = s2n_recv(conn, buffer + bytes_read, sizeof(buffer) - bytes_read, &blocked);
+        if (r == 0) {
+            break;
+        } else if (r > 0) {
+            bytes_read += r;
+        } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            printf("Error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return &test_failure;
+        }
+    }
+
+    if (bytes_read != sizeof(buffer)) {
+        printf("Wrong number of bytes read: %i / %li\n", bytes_read, sizeof(buffer));
+        return &test_failure;
+    }
+
+    if (memcmp(data_to_send, buffer, bytes_read) != 0) {
+        printf("Wrong bytes read: %.*s\n", bytes_read, buffer);
+        return &test_failure;
+    }
+
+    return &test_success;
+}
+
+/* Should match the s2n_shutdown example in USAGE-GUIDE.md */
+static S2N_RESULT s2n_test_shutdown(void *conn)
+{
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    while (s2n_shutdown(conn, &blocked) != S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            printf("Error: %s. %s\n", s2n_strerror(s2n_errno, NULL), s2n_strerror_debug(s2n_errno, NULL));
+            return S2N_RESULT_ERROR;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_test_application(struct s2n_connection *conn)
+{
+    RESULT_GUARD(s2n_test_negotiate(conn));
+
+    /* For testing, we force frequent KeyUpdates */
+    struct s2n_record_algorithm record_alg = *conn->secure->cipher_suite->record_alg;
+    record_alg.encryption_limit = 3;
+    conn->secure->cipher_suite->record_alg = &record_alg;
```

Comment:
Are we actually sending keyupdates in this test? How many records does 50000 bytes translate to?